File: core/shared/src/test/scala/diplomacy/variant/standard/rule/RetreatResolverSpec.scala
package diplomacy.variant.standard.rule

import diplomacy.UnitSpec
import diplomacy.board._
import diplomacy.rule.{ OrderResult, InvalidOrderMessage }
import diplomacy.variant.standard.map
import diplomacy.variant.standard.map.Power
import diplomacy.variant.standard.map.Keywords._
import diplomacy.variant.standard.board.{ Turn => T }
import diplomacy.variant.standard.board.Keywords._
import diplomacy.variant.standard.board.Implicits._
import diplomacy.variant.standard.rule.Keywords._
import diplomacy.variant.standard.rule.Implicits._

class RetreatResolverSpec extends UnitSpec with UsesResolvedResult {
  val resolver = new RetreatResolver {
    type Turn = T
    type Power = map.Power
  }
  val board = {
    new Board[T, map.Power](
      map.map,
      1901.Spring - Retreat,
      Set(
        DiplomacyUnit(France, Army, Bur),
        DiplomacyUnit(France, Army, Mar),
        DiplomacyUnit(Italy, Fleet, Wes)
      ),
      Map(
        DiplomacyUnit(France, Army, Mar) -> UnitStatus.Dislodged(Gas.province),
        DiplomacyUnit(Italy, Fleet, Wes) -> UnitStatus.Dislodged(Tyn.province)
      ),
      Map(Pie.province -> ProvinceStatus(None, true))
    )
  }
  val $ = StandardRuleOrderHelper(board)
  import $._

  "A retreat-resolver" can {
    "resolve a disband order." in {
      val Right(result) =
        resolver(_.copy(season = Autumn))(board, Set(A(Mar).disband(), F(Wes).disband()))
      result.result should be(
        Set(
          new Executed(A(Mar).disband(), Result.Success),
          new Executed(F(Wes).disband(), Result.Success)
        )
      )
      result.board should be(
        board.copy(
          state = 1901.Autumn - Movement,
          units = Set(DiplomacyUnit(France, Army, Bur)),
          unitStatuses = Map[DiplomacyUnit, UnitStatus](),
          provinceStatuses = Map[Province, ProvinceStatus]()
        )
      )
    }
    "resolve a retreat order (1)." in {
      val Right(result) =
        resolver(_.copy(season = Autumn))(board, Set(A(Mar).retreat(Spa), F(Wes).retreat(NAf)))
      result.result should be(
        Set(
          new Executed(A(Mar).retreat(Spa), Result.Success),
          new Executed(F(Wes).retreat(NAf), Result.Success)
        )
      )
      result.board should be(
        board.copy(
          state = 1901.Autumn - Movement,
          units = Set(
            DiplomacyUnit(France, Army, Bur),
            DiplomacyUnit(France, Army, Spa),
            DiplomacyUnit(Italy, Fleet, NAf)
          ),
          unitStatuses = Map[DiplomacyUnit, UnitStatus](),
          provinceStatuses = Map[Province, ProvinceStatus]()
        )
      )
    }
    "resolve a retreat order (2)." in {
      val Right(result) =
        resolver(_.copy(season = Autumn))(board, Set(A(Mar).retreat(Spa), F(Wes).retreat(Spa)))
      result.result should be(
        Set(
          new Executed(A(Mar).retreat(Spa), Result.Failed),
          new Executed(F(Wes).retreat(Spa), Result.Failed)
        )
      )
      result.board should be(
        board.copy(
          state = 1901.Autumn - Movement,
          units = Set(
            DiplomacyUnit(France, Army, Bur)
          ),
          unitStatuses = Map[DiplomacyUnit, UnitStatus](),
          provinceStatuses = Map[Province, ProvinceStatus]()
        )
      )
    }
  }
  it when {
    "resolves a buildable turn" should {
      "go to Build phase." in {
        val board = {
          Board[State, Power, MilitaryBranch, UnitStatus, ProvinceStatus](
            map.map,
            1901.Autumn - Retreat,
            Set(
              DiplomacyUnit(France, Army, Bur),
              DiplomacyUnit(France, Army, Mar),
              DiplomacyUnit(Italy, Fleet, Wes)
            ), Map(), Map()
          )
        }
        val $ = StandardRuleOrderHelper(board)
        import $._

        val Right(result) = resolver(_.copy(season = Autumn))(board, Set())
        result.board should be(
          board.copy(
            state = 1901.Autumn - Build,
            provinceStatuses = Map(
              Bur.province -> ProvinceStatus(Option(France), false),
              Mar.province -> ProvinceStatus(Option(France), false),
              Wes.province -> ProvinceStatus(Option(Italy), false)
            )
          ))
      }
    }
    "a power controls half of supply centers" should {
      "finish the game." in {
        val board = {
          Board[State, Power, MilitaryBranch, UnitStatus, ProvinceStatus](
            map.map, 1901.Autumn - Retreat,
            Set(),
            Map(),
            Map(
              StP.province -> ProvinceStatus[Power](Option(France), false),
              Swe.province -> ProvinceStatus[Power](Option(France), false),
              Nwy.province -> ProvinceStatus[Power](Option(France), false),
              Den.province -> ProvinceStatus[Power](Option(France), false),
              Lvp.province -> ProvinceStatus[Power](Option(France), false),
              Edi.province -> ProvinceStatus[Power](Option(France), false),
              Lon.province -> ProvinceStatus[Power](Option(France), false),
              Ber.province -> ProvinceStatus[Power](Option(France), false),
              Kie.province -> ProvinceStatus[Power](Option(France), false),
              Mun.province -> ProvinceStatus[Power](Option(France), false),
              Hol.province -> ProvinceStatus[Power](Option(France), false),
              Bel.province -> ProvinceStatus[Power](Option(France), false),
              Bre.province -> ProvinceStatus[Power](Option(France), false),
              Par.province -> ProvinceStatus[Power](Option(France), false),
              Mar.province -> ProvinceStatus[Power](Option(France), false),
              Spa.province -> ProvinceStatus[Power](Option(France), false),
              Por.province -> ProvinceStatus[Power](Option(France), false),
              Tun.province -> ProvinceStatus[Power](Option(France), false)
            )
          )
        }
        val $ = StandardRuleOrderHelper(board)
        import $._

        val Right(result) = resolver(_.copy(season = Autumn))(board, Set())
        result.isFinished should be(true)
      }
    }
  }
}

File: test/variant/standard/rule/retreat-resolver-spec.ts
import * as chai from "chai"
import {
  RetreatResolver, Board, Unit, MilitaryBranch, State, Dislodged, Result, Phase, ProvinceStatus,
  StandardRuleHelper as Helper
} from "../../../../src/variant/standard/rule/module"
import { Executed } from "./../../../../src/rule/module"
import { locations as $, Power, map } from "./../../../../src/variant/standard/map/module"
import { Season, Turn } from "./../../../../src/variant/standard/board/module"

const { Army, Fleet } = MilitaryBranch
const { Build, Retreat, Movement } = Phase

const should = chai.should()

const resolver = new RetreatResolver()

describe("A RetreatResolver", () => {
  const u1 = new Unit(Army, $.Mar, Power.Germany)
  const u2 = new Unit(Fleet, $.Wes, Power.Italy)
  const board = new Board<Power>(
    map, new State(new Turn(1901, Season.Spring), Retreat),
    [
      new Unit(Army, $.Bur, Power.Germany),
      u1,
      u2
    ],
    [
      [u1, new Dislodged($.Gas.province)], [u2, new Dislodged($.Tyn.province)]
    ],
    [[$.Pie.province, new ProvinceStatus(null, true)]]
  )
  const $$ = new Helper(board)
  describe("A RetreatResolver", () => {
    it("resolves a disband order", () => {
      const result =
        resolver.resolve(board, new Set([$$.A($.Mar).disband(), $$.F($.Wes).disband()])).result
      Array.from(result.results).should.have.deep.members([
        new Executed($$.A($.Mar).disband(), Result.Success),
        new Executed($$.F($.Wes).disband(), Result.Success)
      ])
      result.board.should.deep.equal(new Board(
        map, new State(new Turn(1901, Season.Autumn), Movement),
        [new Unit(Army, $.Bur, Power.Germany)],
        [], []
      ))
    })
    it("resolves a retreat order (1)", () => {
      const result = resolver.resolve(
        board, new Set([$$.A($.Mar).retreat($.Spa), $$.F($.Wes).retreat($.NAf)])
      ).result
      Array.from(result.results).should.have.deep.members([
        new Executed($$.A($.Mar).retreat($.Spa), Result.Success),
        new Executed($$.F($.Wes).retreat($.NAf), Result.Success)
      ])
      result.board.should.deep.equal(new Board(
        map, new State(new Turn(1901, Season.Autumn), Movement),
        [
          new Unit(Army, $.Bur, Power.Germany),
          new Unit(Army, $.Spa, Power.Germany),
          new Unit(Fleet, $.NAf, Power.Italy)
        ],
        [], []
      ))
    })
    it("resolves a retreat order (2)", () => {
      const result = resolver.resolve(
        board, new Set([$$.A($.Mar).retreat($.Spa), $$.F($.Wes).retreat($.Spa)])
      ).result
      Array.from(result.results).should.have.deep.members([
        new Executed($$.A($.Mar).retreat($.Spa), Result.Failed),
        new Executed($$.F($.Wes).retreat($.Spa), Result.Failed)
      ])
      result.board.should.deep.equal(new Board(
        map, new State(new Turn(1901, Season.Autumn), Movement),
        [new Unit(Army, $.Bur, Power.Germany)],
        [], []
      ))
    })
  })
  describe("when resolves a buildable turn", () => {
    it("go to Build phase", () => {
      const board = new Board<Power>(
        map, new State(new Turn(1901, Season.Autumn), Retreat),
        [], [], []
      )
      const result = resolver.resolve(board, new Set()).result
      result.board.state.should.deep.equal(new State(new Turn(1901, Season.Autumn), Build))
    })
  })
  describe("when a power controls half of supply centers", () => {
    it("finish the game", () => {
      const board = new Board<Power>(
        map, new State(new Turn(1901, Season.Autumn), Retreat),
        [], [],
        [
          [$.StP.province, new ProvinceStatus(Power.France, false)],
          [$.Swe.province, new ProvinceStatus(Power.France, false)],
          [$.Nwy.province, new ProvinceStatus(Power.France, false)],
          [$.Den.province, new ProvinceStatus(Power.France, false)],
          [$.Lvp.province, new ProvinceStatus(Power.France, false)],
          [$.Edi.province, new ProvinceStatus(Power.France, false)],
          [$.Lon.province, new ProvinceStatus(Power.France, false)],
          [$.Ber.province, new ProvinceStatus(Power.France, false)],
          [$.Kie.province, new ProvinceStatus(Power.France, false)],
          [$.Mun.province, new ProvinceStatus(Power.France, false)],
          [$.Hol.province, new ProvinceStatus(Power.France, false)],
          [$.Bel.province, new ProvinceStatus(Power.France, false)],
          [$.Bre.province, new ProvinceStatus(Power.France, false)],
          [$.Par.province, new ProvinceStatus(Power.France, false)],
          [$.Mar.province, new ProvinceStatus(Power.France, false)],
          [$.Spa.province, new ProvinceStatus(Power.France, false)],
          [$.Por.province, new ProvinceStatus(Power.France, false)],
          [$.Tun.province, new ProvinceStatus(Power.France, false)]
        ]
      )
      const result = resolver.resolve(board, new Set()).result
      result.isFinished.should.equal(true)
    })
  })
})
